Phase-1 blog post for GitHub Checks API Plugin (#3499)
jenkins-infra/jenkins.io: content/blog/2020/07/2020-07-03-github-checks-api-plugin-coding-phase-1.adoc

---
layout: post
title: "GitHub Checks API Plugin Project - Coding Phase 1"
tags:
- github
- api
- plugins
- gsoc
- gsoc2020
author: XiongKezhi
opengraph:
  image: /images/gsoc/opengraph.png
---

== Introduction

This blog post is about our coding phase 1 progress on GSoC project: link:/projects/gsoc/2020/projects/github-checks/[GitHub Checks API Plugin].

> In general, the GitHub Checks API is a highly customized way to integrate CI tools to make reports for PRs. 

Reviewer: ```suggestion

The GitHub Checks API is a highly customized way to integrate CI tools to make reports for PRs. 
```

It allows users to see CI reports on GitHub pages directly.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-run.png[GitHub Checks API]

What's more exciting is that it can leave annotations on specific lines of code, just as the comments people left while reviewing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-annotations.png[Annotations]

While on Jenkins' side, the source code view provided by link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] does pretty much the same thing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/source-view.png[Source Code View]

So if we could utilize such features through GitHub Checks API, it would make Jenkins more convenient to GitHub users.

== Features from Coding Phase 1

In the past month, our team was mostly working on the general checks API and an implementation for GitHub checks API.

=== General Checks API

> Although the general checks API is developed based on the semantic meaning of GitHub Checks API, we still want to prepare it for similar concepts on other platforms like link:https://docs.gitlab.com/ee/api/commits.html#commit-status[Commit Status API] from GitLab. Contributions for implementations on these platforms are always welcomed!

Reviewer: One sentence per line makes it easier to review and see diffs.
```suggestion
Although the general checks API is developed based on the semantic meaning of GitHub Checks API, we still want to prepare it for similar concepts on other platforms like link:https://docs.gitlab.com/ee/api/commits.html#commit-status[Commit Status API] from GitLab.
Contributions for implementations on these platforms are always welcomed!
```

one question though are they always welcomed? What about welcomed in the future? Do we want the distraction right now?

Author: ok, we can leave this until the last blog post, I didn't expect someone would contribute to it now anyway :smile:

                                .withStartColumn(0)
                                .withEndColumn(20)
                                .withTitle("Hello Jenkins")
                                .withRawDetails("a simple echo command")
                                .build(),
                        new ChecksAnnotationBuilder()
                                .withPath("Jenkinsfile")
                                .withLine(2)
                                .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
                                .withMessage("say hello to GitHub Checks API")
                                .withStartColumn(0)
                                .withEndColumn(30)
                                .withTitle("Hello GitHub Checks API")
                                .withRawDetails("a simple echo command")
                                .build()))
                .build())
        .withActions(Collections.singletonList(
                new ChecksAction("re-run", "re-run Jenkins build", "#0")))
        .build();

ChecksPublisher publisher = ChecksPublisherFactory.fromRun(run);
publisher.publish(details);

----

== Future Works

> As mentioned in <<Introduction>>, utilizing some features from other plugins can improve the Jenkins user experience on GitHub. Thus, our next step is letting link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] and link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin] consume our API. Besides, pipeline support is in our plan as well.

Reviewer: ```suggestion
Our next step is letting link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] and link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin] consume our API. Besides, pipeline support is in our plan as well.
```

What's more exciting is that it can leave annotations on specific lines of code, just as the comments people left while reviewing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-annotations.png[Annotations]

While on Jenkins' side, the source code view provided by link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] does pretty much the same thing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/source-view.png[Source Code View]

So if we could utilize such features through GitHub Checks API, it would make Jenkins more convenient to GitHub users.

== Features from Coding Phase 1

In the past month, our team was mostly working on the general checks API and an implementation for GitHub checks API.

=== General Checks API

Although the general checks API is developed based on the semantic meaning of GitHub Checks API, we still want to prepare it for similar concepts on other platforms like link:https://docs.gitlab.com/ee/api/commits.html#commit-status[Commit Status API] from GitLab.

> Contributions for implementations on these platforms are always welcomed!

Reviewer: ```suggestion
Contributions for implementations on these platforms will be welcomed in the future.
```

                                .withLine(1)
                                .withAnnotationLevel(ChecksAnnotationLevel.NOTICE)
                                .withMessage("say hello to Jenkins")
                                .withStartColumn(0)
                                .withEndColumn(20)
                                .withTitle("Hello Jenkins")
                                .withRawDetails("a simple echo command")
                                .build(),
                        new ChecksAnnotationBuilder()
                                .withPath("Jenkinsfile")
                                .withLine(2)
                                .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
                                .withMessage("say hello to GitHub Checks API")
                                .withStartColumn(0)
                                .withEndColumn(30)
                                .withTitle("Hello GitHub Checks API")
                                .withRawDetails("a simple echo command")
                                .build()))
                .build())
        .withActions(Collections.singletonList(

>                 new ChecksAction("re-run", "re-run Jenkins build", "#0")))

Reviewer: isn't re-run built into the API? is it a realistic action?

Author: Just an example here, the action is not supported yet. I'll change it to something else to avoid misconceptions.

---
layout: post
title: "GitHub Checks API Plugin Project - Coding Phase 1"
tags:
- github
- api

> - plugins

Reviewer 2: ```suggestion
- plugins
- developer
```

- gsoc
- gsoc2020
author: XiongKezhi
opengraph:
  image: /images/gsoc/opengraph.png
---

> == Introduction

Reviewer 2: ```suggestion
```

Not needed

This blog post is about our coding phase 1 progress on GSoC project: link:/projects/gsoc/2020/projects/github-checks/[GitHub Checks API Plugin].

> The GitHub Checks API is a highly customized way to integrate CI tools to make reports for PRs. 

Reviewer 2: ```suggestion
The GitHub Checks API is a highly customized way to integrate CI tools to make reports for pull-requests (PRs). 
```

It allows users to see CI reports on GitHub pages directly.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-run.png[GitHub Checks API]

What's more exciting is that it can leave annotations on specific lines of code, just as the comments people left while reviewing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-annotations.png[Annotations]

While on Jenkins' side, the source code view provided by link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] does pretty much the same thing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/source-view.png[Source Code View]

> So if we could utilize such features through GitHub Checks API, it would make Jenkins more convenient to GitHub users.

Reviewer 3: I would change “So if we could” to just start with “Utilizing”

== Features from Coding Phase 1

In the past month, our team was mostly working on the general checks API and an implementation for GitHub checks API.

=== General Checks API

Although the general checks API is developed based on the semantic meaning of GitHub Checks API, we still want to prepare it for similar concepts on other platforms like link:https://docs.gitlab.com/ee/api/commits.html#commit-status[Commit Status API] from GitLab.
Contributions for implementations on these platforms will be welcomed in the future.

=== GitHub Checks API Implmentation

> Since we now have implemented the support for GitHub Checks API, plugin developers can easily create a GitHub check run for a GitHub branch source project by consuming our API.

Reviewer 2: They cannot, because this functionality has not been released yet. Please either update guidelines or describe how to use it (required for alpha releases as well)

**Example:** To create a check run like:

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/created-check-run.png[Created Check Run]

> Consumers just need to use our API in this way:

Reviewer 2: The API below is quite complicated, so probably "just" is not appropriate

                                .withStartColumn(0)
                                .withEndColumn(20)
                                .withTitle("Hello Jenkins")
                                .withRawDetails("a simple echo command")
                                .build(),
                        new ChecksAnnotationBuilder()
                                .withPath("Jenkinsfile")
                                .withLine(2)
                                .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
                                .withMessage("say hello to GitHub Checks API")
                                .withStartColumn(0)
                                .withEndColumn(30)
                                .withTitle("Hello GitHub Checks API")
                                .withRawDetails("a simple echo command")
                                .build()))
                .build())
        .withActions(Collections.singletonList(
                new ChecksAction("formatting", "format code", "#0")))
        .build();

ChecksPublisher publisher = ChecksPublisherFactory.fromRun(run);
publisher.publish(details);

----

> 

Reviewer 2: Add a section with a demo from Thursday? You can embed video into the blogpost

== Future Works

> Our next step is letting link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] and link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin] consume our API.

Reviewer: ```suggestion
Our next step is integrating our API into link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] and link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin].
```

Pipeline support is in our plan as well.

> == Resources and link

Reviewer: ```suggestion
== Resources
```

---
layout: post
title: "GitHub Checks API Plugin Project - Coding Phase 1"
tags:
- github
- api
- plugins
- developer
- gsoc
- gsoc2020
author: XiongKezhi
opengraph:
  image: /images/gsoc/opengraph.png
---

This blog post is about our coding phase 1 progress on GSoC project: link:/projects/gsoc/2020/projects/github-checks/[GitHub Checks API Plugin].

The GitHub Checks API is a highly customized way to integrate CI tools to make reports for pull-requests (PRs). 
It allows users to see CI reports on GitHub pages directly.

> image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-run.png[GitHub Checks API]

Reviewer 4: For all images in this blog post: if these are copied from GitHub help pages (i.e. not your screenshots) then you should correctly cite them.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/github-check-annotations.png[Annotations]

While on Jenkins' side, the source code view provided by link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] does pretty much the same thing.

image:/images/post-images/2020/07-github-checks-api-plugin-coding-phase-1/source-view.png[Source Code View]

Utilizing such features through GitHub Checks API, it would make Jenkins more convenient to GitHub users.

== Features from Coding Phase 1

In the past month, our team was mostly working on the general checks API and an implementation for GitHub checks API.

.GitHub Checks API Plugin Demo [starts from 50:15]
video::HQLhakpx5mk[youtube,width=800,height=420]

=== General Checks API

Although the general checks API is developed based on the semantic meaning of GitHub Checks API, we still want to prepare it for similar concepts on other platforms like link:https://docs.gitlab.com/ee/api/commits.html#commit-status[Commit Status API] from GitLab.
Contributions for implementations on these platforms will be welcomed in the future.

> === GitHub Checks API Implmentation

Reviewer 4: ```suggestion
=== GitHub Checks API Implementation
```

> === GitHub Checks API Implmentation

Reviewer 4: Maybe it makes sense to write that checks are already running even if no consumer has been implemented yet.

> Our work on supporting GitHub Checks API are mostly done by now, so after the releasing, Jenkins developers can then create GitHub checks for a GitHub branch source project by consuming our API.

Reviewer 4: ```suggestion
Our work on supporting GitHub Checks API is mostly done by now, so after the releasing, Jenkins developers can then create GitHub checks for a GitHub branch source project by consuming our API.
```

                                .build(),
                        new ChecksAnnotationBuilder()
                                .withPath("Jenkinsfile")
                                .withLine(2)
                                .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
                                .withMessage("say hello to GitHub Checks API")
                                .withStartColumn(0)
                                .withEndColumn(30)
                                .withTitle("Hello GitHub Checks API")
                                .withRawDetails("a simple echo command")
                                .build()))
                .build())
        .withActions(Collections.singletonList(
                new ChecksAction("formatting", "format code", "#0")))
        .build();

ChecksPublisher publisher = ChecksPublisherFactory.fromRun(run);
publisher.publish(details);

----

== Future Works

The next step is integrating our API into link:https://plugins.jenkins.io/warnings-ng/[Warnings Next Generation Plugin] and link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin] consume our API.

> Pipeline support is in our plan as well.

Reviewer 4: We should make clear what Pipeline support means: that checks can be created and updated directly in a pipeline script (without requiring a consumer plugin that support the checks).
